Manage CacheStorage by origin

szewai opened the pull request:
#### 03403b4f8c4e7aa6e10be71dca42f10c0f28bbdd
<pre>
Manage CacheStorage by origin
<a href="https://bugs.webkit.org/show_bug.cgi?id=247207">https://bugs.webkit.org/show_bug.cgi?id=247207</a>
rdar://problem/101968609

Reviewed by Youenn Fablet.

The OriginStorageManager class is designed to manage storage of a paritioned origin, and have exclusive access to origin
storage directory. This patch refactors DOMCache backend to use OriginStorageManager for CacheStorage management.

The old structure is:
NetworkSession - CacheStorage::Engine - CacheStorage::Caches - CacheStorage::Cache
The new structure is:
NetworkSession - NetworkStorageManager - OriginStorageManager - CacheStorageManager - CacheStorageCache

Here are the new classes:
CacheStorageManager: it manages cache storage of a partitioned origin and store caches.
CacheStorageCache: it represents a cache and keeps track of records (by record information) in memory.
CacheStorageStore: it represents backing store of cache; it is implemented by CacheStorageMemoryStore and
CacheStorageDiskStore.
CacheStorageRegistry: it keeps track of all CacheStorageCaches in a session, so NetworkStorageManager can find the
target of a CacheStorage IPC message quickly.

Here are some other changes made to complete the transition:
1. CacheStorage messages are sent to NetworkStorageManager instead of CacheStorage::Engine.
2. NetworkStorageManager handles CachesStorage messages and performs corresponding operations on its own WorkQueue.
Since WorkQueue is not tied to one thread, this patch use ResourceResponse::CrossThreadData instead of ResourceResponse
in CacheStorageRecord.
3. CacheStoragePersistentStore provides a simple file system backend for CacheStorage, instead of using
NetworkCache::Storage.
4. Representation string of CacheStorage will report origins that have CacheStorage data. It used to report origins of
existing CacheStorage::Caches (which manages cache storage of partitioned origin like CacheStorageManager), but having
CacheStorage::Caches does not mean the origin has CacheStorage data or the origin is actively performing CacheStorage
operations, and the state is subjective to change based on implementation (e.g. we might add/delete CacheStorage::Caches
proactively based on its activity). Therefore, let&apos;s use CacheStorage data as the truth, and update test expectations.
5. CacheStorage backend used to have multiple WorkQueues per partition origin, now it has one WorkQueue for all origins
(shared with other types) for message handling, and at most one queue per cache for I/O operation on cache records.

* LayoutTests/http/tests/cache-storage/cache-clearing-origin.https.html:
* LayoutTests/http/tests/cache-storage/cache-origins.https.html:
* LayoutTests/http/tests/cache-storage/cache-representation.https.html:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::overwriteEntireFile):
* Source/WTF/wtf/FileSystem.h:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::CrossThreadData::copy const):
* Source/WebCore/platform/network/ResourceResponseBase.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteDataForOrigins):
(WebKit::NetworkProcess::deleteAndRestrictWebsiteDataForRegistrableDomains):
(WebKit::NetworkProcess::registrableDomainsWithWebsiteData):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp: Added.
(WebKit::computeKeyURL):
(WebKit::nextRecordIdentifier):
(WebKit::createStore):
(WebKit::CacheStorageCache::CacheStorageCache):
(WebKit::CacheStorageCache::manager):
(WebKit::CacheStorageCache::getSize):
(WebKit::CacheStorageCache::open):
(WebKit::toCacheStorageRecord):
(WebKit::CacheStorageCache::retrieveRecords):
(WebKit::CacheStorageCache::removeRecords):
(WebKit::CacheStorageCache::findExistingRecord):
(WebKit::CacheStorageCache::putRecords):
(WebKit::CacheStorageCache::putRecordsAfterQuotaCheck):
(WebKit::CacheStorageCache::putRecordsInStore):
(WebKit::CacheStorageCache::removeAllRecords):
(WebKit::CacheStorageCache::close):
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.h: Added.
(WebKit::CacheStorageCache::identifier const):
(WebKit::CacheStorageCache::name const):
(WebKit::CacheStorageCache::uniqueName const):
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp: Added.
(WebKit::shouldStoreBodyAsBlob):
(WebKit::computeSHA1):
(WebKit::RecordMetaData::RecordMetaData):
(WebKit::CacheStorageDiskStore::create):
(WebKit::CacheStorageDiskStore::CacheStorageDiskStore):
(WebKit::CacheStorageDiskStore::versionDirectoryPath const):
(WebKit::CacheStorageDiskStore::saltFilePath const):
(WebKit::CacheStorageDiskStore::recordsDirectoryPath const):
(WebKit::CacheStorageDiskStore::recordFilePath const):
(WebKit::CacheStorageDiskStore::recordBlobFilePath const):
(WebKit::CacheStorageDiskStore::blobsDirectoryPath const):
(WebKit::CacheStorageDiskStore::blobFilePath const):
(WebKit::decodeRecordMetaData):
(WebKit::decodeRecordHeader):
(WebKit::CacheStorageDiskStore::readRecordFromFileData):
(WebKit::CacheStorageDiskStore::readAllRecords):
(WebKit::CacheStorageDiskStore::readRecords):
(WebKit::CacheStorageDiskStore::deleteRecords):
(WebKit::encodeRecordHeader):
(WebKit::encodeRecordBody):
(WebKit::encodeRecord):
(WebKit::CacheStorageDiskStore::writeRecords):
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.h: Added.
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp: Added.
(WebKit::nextUpdateNumber):
(WebKit::readCachesList):
(WebKit::writeCachesList):
(WebKit::readSizeFile):
(WebKit::writeSizeFile):
(WebKit::CacheStorageManager::cacheStorageSalt):
(WebKit::CacheStorageManager::cacheStorageOriginDirectory):
(WebKit::CacheStorageManager::originsOfCacheStorageData):
(WebKit::getDirectorySize):
(WebKit::CacheStorageManager::cacheStorageSize):
(WebKit::CacheStorageManager::hasCacheList):
(WebKit::CacheStorageManager::makeDirty):
(WebKit::CacheStorageManager::CacheStorageManager):
(WebKit::CacheStorageManager::~CacheStorageManager):
(WebKit::CacheStorageManager::initializeCaches):
(WebKit::CacheStorageManager::openCache):
(WebKit::CacheStorageManager::removeCache):
(WebKit::CacheStorageManager::allCaches):
(WebKit::CacheStorageManager::initializeCacheSize):
(WebKit::CacheStorageManager::finishInitializingSize):
(WebKit::CacheStorageManager::requestSpaceAfterInitializingSize):
(WebKit::CacheStorageManager::requestSpace):
(WebKit::CacheStorageManager::sizeIncreased):
(WebKit::CacheStorageManager::sizeDecreased):
(WebKit::CacheStorageManager::reference):
(WebKit::CacheStorageManager::dereference):
(WebKit::CacheStorageManager::connectionClosed):
(WebKit::CacheStorageManager::removeUnusedCache):
(WebKit::CacheStorageManager::hasDataInMemory):
(WebKit::CacheStorageManager::isActive):
(WebKit::CacheStorageManager::representationString):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.h: Added.
(WebKit::CacheStorageManager::salt const):
* Source/WebKit/NetworkProcess/storage/CacheStorageMemoryStore.cpp: Added.
(WebKit::CacheStorageMemoryStore::create):
(WebKit::CacheStorageMemoryStore::readAllRecords):
(WebKit::CacheStorageMemoryStore::readRecords):
(WebKit::CacheStorageMemoryStore::deleteRecords):
(WebKit::CacheStorageMemoryStore::writeRecords):
* Source/WebKit/NetworkProcess/storage/CacheStorageMemoryStore.h: Added.
* Source/WebKit/NetworkProcess/storage/CacheStorageRecord.h: Added.
(WebKit::CacheStorageRecordInformation::updateVaryHeaders):
(WebKit::CacheStorageRecord::CacheStorageRecord):
(WebKit::CacheStorageRecord::copy const):
* Source/WebKit/NetworkProcess/storage/CacheStorageRegistry.cpp: Added.
(WebKit::CacheStorageRegistry::registerCache):
(WebKit::CacheStorageRegistry::unregisterCache):
(WebKit::CacheStorageRegistry::cache):
* Source/WebKit/NetworkProcess/storage/CacheStorageRegistry.h: Added.
* Source/WebKit/NetworkProcess/storage/CacheStorageStore.h: Added.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::canHandleTypes):
(WebKit::NetworkStorageManager::originStorageManager):
(WebKit::NetworkStorageManager::getAllOrigins):
(WebKit::NetworkStorageManager::cacheStorageOpenCache):
(WebKit::NetworkStorageManager::cacheStorageRemoveCache):
(WebKit::NetworkStorageManager::cacheStorageAllCaches):
(WebKit::NetworkStorageManager::cacheStorageReference):
(WebKit::NetworkStorageManager::cacheStorageDereference):
(WebKit::NetworkStorageManager::cacheStorageRetrieveRecords):
(WebKit::NetworkStorageManager::cacheStorageRemoveRecords):
(WebKit::NetworkStorageManager::cacheStoragePutRecords):
(WebKit::NetworkStorageManager::cacheStorageClearMemoryRepresentation):
(WebKit::NetworkStorageManager::cacheStorageRepresentation):
(WebKit::readOriginFromFile): Deleted.
(WebKit::writeOriginToFile): Deleted.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::existingCacheStorageManager):
(WebKit::OriginStorageManager::StorageBucket::StorageBucket):
(WebKit::OriginStorageManager::StorageBucket::connectionClosed):
(WebKit::OriginStorageManager::StorageBucket::toStorageType const):
(WebKit::OriginStorageManager::StorageBucket::toStorageIdentifier const):
(WebKit::OriginStorageManager::StorageBucket::cacheStorageManager):
(WebKit::OriginStorageManager::StorageBucket::isActive const):
(WebKit::OriginStorageManager::StorageBucket::isEmpty):
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInListFromMemory):
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInListFromDisk):
(WebKit::OriginStorageManager::StorageBucket::deleteData):
(WebKit::OriginStorageManager::StorageBucket::deleteCacheStorageData):
(WebKit::OriginStorageManager::StorageBucket::deleteEmptyDirectory):
(WebKit::OriginStorageManager::StorageBucket::resolvedPath):
(WebKit::OriginStorageManager::StorageBucket::closeCacheStorageManager):
(WebKit::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
(WebKit::OriginStorageManager::defaultBucket):
(WebKit::OriginStorageManager::existingCacheStorageManager):
(WebKit::OriginStorageManager::cacheStorageManager):
(WebKit::OriginStorageManager::closeCacheStorageManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/NetworkProcess/storage/StorageUtilities.h: Added.
(WebKit::readOriginFromFile):
(WebKit::writeOriginToFile):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::open):
(WebKit::WebCacheStorageConnection::remove):
(WebKit::WebCacheStorageConnection::retrieveCaches):
(WebKit::WebCacheStorageConnection::retrieveRecords):
(WebKit::WebCacheStorageConnection::batchDeleteOperation):
(WebKit::WebCacheStorageConnection::batchPutOperation):
(WebKit::WebCacheStorageConnection::reference):
(WebKit::WebCacheStorageConnection::dereference):
(WebKit::WebCacheStorageConnection::clearMemoryRepresentation):
(WebKit::WebCacheStorageConnection::engineRepresentation):

Canonical link: <a href="https://commits.webkit.org/258235@main">https://commits.webkit.org/258235@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ffe22b2ab5258c9aa7d53925aa741d77499fb3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101212 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10366 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110503 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170784 "Found 1 new test failure: js/dom/Promise-reject-large-string.html (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105200 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11302 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1233 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93617 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108328 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106995 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8602 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91845 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35147 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90503 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23251 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78151 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/91665 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4007 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24765 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/87784 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/1575 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4057 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1191 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29283 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10158 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44240 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/90679 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5667 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5811 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20270 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on LayoutTests/http/tests/cache-storage/cache-clearing-origin.https.html:
```diff
@@ -36,17 +36,15 @@
 }, "Create a cache storage from localhost and clear it");
 
 promise_test(async test => {
+    console.log(window.location.origin);
```
probably not needed

szewai replied:
ah right; will remove

youennf commented on Source/WebCore/Modules/fetch/FetchResponse.cpp:
```diff
@@ -131,12 +131,12 @@ ExceptionOr<Ref<FetchResponse>> FetchResponse::create(ScriptExecutionContext& co
 
     r->m_internalResponse.setHTTPStatusCode(status);
-    r->m_internalResponse.setHTTPStatusText(statusText);
+    r->m_internalResponse.setHTTPStatusText(statusText.releaseString());
```
Are these changes needed?

szewai replied:
The CacheStorage message that contains `ResourceResponse` will be dispatched to WorkQueue (not bounded by one thread), so we cannot use `AtomString`(bounded by one thread) in `ResourceResponse`.

Or do you have other proposal?

youennf replied:
Would it be better to split this from the patch?

szewai replied:
I included it in this patch as it indicates why we want this change; I can split it out if you prefer that

youennf replied:
If it is not too much work, let's do that. It might help bisecting/reverting if anything goes wrong.

youennf replied:
Actually, can we use crossThreadData/fromCrossThreadData instead?

szewai replied:
Yes, I've updated the patch to use `crossThreadData`.

youennf commented on Source/WebCore/Modules/websockets/WebSocketHandshake.cpp:
```diff
@@ -426,7 +426,7 @@ int WebSocketHandshake::readStatusLine(const uint8_t* header, size_t headerLengt
 
     statusCode = parseInteger<int>(statusCodeString).value();
-    statusText = AtomString(space2 + 1, end - space2 - 3); // Exclude "\r\n".
+    statusText = String(space2 + 1, end - space2 - 3); // Exclude "\r\n".
```
Ditto, I do not really understand why there are changes to WebSocket here.

szewai replied:
Ditto.

youennf commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1509,12 +1509,6 @@ void NetworkProcess::fetchWebsiteData(PAL::SessionID sessionID, OptionSet<Websit
-            callbackAggregator->m_websiteData.entries.appendVector(entries);
-        });
-    }
-
```
I am not sure why we are removing the possibility to remove DOMCache only. Can you clarify? Is it because all types handled by the network storage manager can only be removed as a whole?

szewai replied:
Yes, we call `NetworkStorageManager::deleteData` below, which can handle types in `NetworkStorageManager::canHandleTypes`.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,386 @@
+    if (url.isNull()) {
+        for (auto urlRecords : m_records.values()) {
+            for (auto record : urlRecords)
+                targetRecordInfos.append(record);
```
Can we use WTF::map?

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,386 @@
+            return a.identifier < b.identifier;
+        });
+
+        callback(result);
```
WTFMove(result) might reduce a copy

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
```diff
@@ -0,0 +1,500 @@
+    if (!count)
+        return std::nullopt;
+
+    result.reserveCapacity(*count);
```
reserveInitialCapacity.

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
```diff
@@ -0,0 +1,500 @@
+            for (auto& fileName : fileNames) {
+                // Files in /Blobs directory are hard link.
+                if (fileName == "Blobs"_s)
+                    continue;
```
Why is it needed?

szewai replied:
This is needed before we delete the Blobs directory (that we used for blob deduplication and will no longer be used after this patch).

youennf commented on Source/WebKit/NetworkProcess/storage/DiskCacheStorageCacheStore.h:
```diff
@@ -0,0 +1,65 @@
+    String recordsDirectoryPath() const;
+    String recordFilePath(const NetworkCache::Key&) const;
+    String recordBlobFilePath(const String&) const;
+    String blobsDirectoryPath() const;
```
Why do we have a blobsDirectoryPath?

szewai replied:
We no longer need this directory since we won't do blob deduplication in CacheStorage; I added the `blobsDirectoryPath` to use as a reminder for deleting the directory after this patch...
I could remove it from this patch.

youennf commented on Source/WebKit/NetworkProcess/storage/DiskCacheStorageCacheStore.cpp:
```diff
@@ -0,0 +1,523 @@
+        recordBlobDatas.append(WTFMove(bodyData));
+    }
+
+    m_ioQueue->dispatch([this, protectedThis = Ref { *this }, recordFiles = crossThreadCopy(recordFiles), recordDatas = crossThreadCopy(recordDatas), recordBlobDatas, callback = WTFMove(callback)]() mutable {
```
WTFMove(recordFiles), WTFMove(recordDatas), WTFMove(recordBlobDatas)

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/DiskCacheStorageCacheStore.h:
```diff
@@ -0,0 +1,65 @@
+
+namespace WebKit {
+
+class DiskCacheStorageCacheStore final : public CacheStorageCacheStore {
```
The name is difficult to parse, I initially thought it was related to disk cache (aka network cache).
I don't have a good name proposal right now.
We were using namespaces previously to help decipher.

szewai replied:
Is it better to use `CacheStoragePersistentStore` and `CacheStorageEphemeralStore`?

youennf replied:
Yes, that sounds better, or CacheStorageDiskStore/CacheStorageMemoryStore depending on the names for other data types.

szewai replied:
I've updated to `CacheStoragePersistentStore` and `CacheStorageEphemeralStore` in the latest patch; will change to  `CacheStorageDiskStore` and `CacheStorageMemoryStore`.

youennf commented on Source/WebKit/NetworkProcess/storage/DiskCacheStorageCacheStore.h:
```diff
@@ -0,0 +1,65 @@
+    std::optional<CacheStorageRecord> readRecordFromFileData(Vector<uint8_t>&, Vector<uint8_t>&);
+
+    String m_cacheName;
+    String m_path;
```
CacheStorageCacheStore is thread safe ref counted so there are chances that it gets destroyed in any thread.
Are we sure DiskCacheStorageCacheStore is destroyed in the right thread so that m_cacheName unrefing will not not be an issue?
Can we have CacheStorageCacheStore be just refcounted instead?

szewai replied:
I think we could use refcounted instead.

youennf commented on Source/WebKit/NetworkProcess/storage/MemoryCacheStorageCacheStore.cpp:
```diff
@@ -0,0 +1,78 @@
+    for (auto& record : m_records.values())
+        result.append(record->copy());
+
+    callback(WTFMove(result));
```
WTF::map maybe?

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/DiskCacheStorageCacheStore.cpp:
```diff
@@ -0,0 +1,523 @@
+{
+    Vector<String> recordFiles;
+    for (auto recordInfo : recordInfos)
+        recordFiles.append(recordFilePath(recordInfo.key));
```
WTF::map maybe?

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/DiskCacheStorageCacheStore.cpp:
```diff
@@ -0,0 +1,523 @@
+    for (auto recordInfo : recordInfos)
+        recordFiles.append(recordFilePath(recordInfo.key));
+
+    m_ioQueue->dispatch([this, protectedThis = Ref { *this }, recordFiles = crossThreadCopy(recordFiles), callback = WTFMove(callback)]() mutable {
```
WTFMove

szewai replied:
Sure.

youennf reviewed (commented):
Overall direction looks ok.
Given the optimisations done in network cache to be as efficient as possible, maybe there is a way to share more code between the two.
Some comments below, I only need a partial pass though.

youennf commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -573,6 +573,22 @@ std::optional<Vector<uint8_t>> readEntireFile(const String& path)
     return contents;
 }
 
+int overwriteEntireFile(const String& path, const void* data, int length)
```
Could be a Span<const uint8_t> to modernise things

szewai replied:
Sure.

youennf commented on Source/WebCore/platform/network/ResourceResponseBase.h:
```diff
@@ -71,11 +71,6 @@ class ResourceResponseBase {
-        CrossThreadData& operator=(const CrossThreadData&) = delete;
-        CrossThreadData() = default;
-        CrossThreadData(CrossThreadData&&) = default;
-
```
Why removing the deletion of the copy constructor?

szewai replied:
In this patch, `CacheStorageRecord` (that contains `CrossThreadData`) is stored in `CacheStorageMemoryStore`, which returns copy of `CacheStorageRecord` on `readRecords` and `readAllRecords`.
I think there are other places `CacheStorageRecord` will be copied, like std::sort() in `CacheStorageCache::open`.

youennf replied:
If a copy is kept and the cross thread data is sent to another thread, we break the assumption.
Removing the copy constructor protects from such mistakes.
We can probably store ResourceResponse in `CacheStorageMemoryStore `, I don't think the potential perf optimization is worth it.

szewai replied:
`CacheStorageMemoryStore` is going to be accessed from different threads (since it's accessed on WorkQueue), so it's not safe to store `ResourceResponse`.
A possible solution is to store `CrossThreadData` in `CacheStorageMemoryStore`, but every time we need to read it, we take it out, convert it to `ResourceResponse`, and call `crossThreadData()` twice on the `ResourceResponse`  to create two `CrossThreadData` -- one for result and one for storage.
And we do the conversion every time we need a copy of CrossThreadData. Is this better than current approach? If so I will make the change.

youennf replied:
If we store `ResourceResponse` and the only thing we are doing is creating a CrossThreadData from it, this should be safe. Maybe we can store `const ResourceResponse` to help ensuring it is never mutated. 
Or you can add a specific method to clone a CrossThreadData.

szewai replied:
hmm it should be safe but I think `AtomString` can only be read from the thread that creates it?

If we still need to store `CrossThreadData` in `CacheStorageRecord` and mark `CrossThreadData` as non-copyable, then `CacheStorageRecord` will need to be non-copyable and the code might be more complicated, e.g. we will need to add more code like constructor (from moving `CrossThreadData`) and `copy()`. Do you want me to make the change?

szewai replied:
Updated the patch by keeping `CrossThreadData` non-copyable.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+}
+
+CacheStorageCache::CacheStorageCache(CacheStorageManager& manager, const String& name, const String& uniqueName, const String& path, Ref<WorkQueue>&& queue)
+    : m_manager(&manager)
```
s/&manager/manager

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+    return keyURL.string();
+}
+
+static uint64_t nextRecordIdentifier()
```
Should we instead introduce a RecordIdentifier?

szewai replied:
I think we could but currently DOMCacheEngine and messages are using `uint64_t`, so we might do that in a followup (since this patch is big).

szewai replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=249591.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+
+CacheStorageRecord CacheStorageCache::toCacheStorageRecord(WebCore::DOMCacheEngine::Record&& record)
+{
+    ASSERT(m_manager);
```
I am not sure this is useful given we call m_manager->salt() just after.

szewai replied:
I will change `toCacheStorageRecord` to a static function.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+            WebCore::DOMCacheEngine::Record record { cacheStorageRecord->info.identifier, 0, cacheStorageRecord->requestHeadersGuard, cacheStorageRecord->request, cacheStorageRecord->options, cacheStorageRecord->referrer, cacheStorageRecord->responseHeadersGuard, { }, nullptr, 0 };
+            if (options.shouldProvideResponse) {
+                record.response = WebCore::ResourceResponse::fromCrossThreadData(WTFMove(cacheStorageRecord->responseData));
+                record.responseBody = cacheStorageRecord->responseBody;
```
WTFMove(cacheStorageRecord->responseBody)

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+                    return callback(makeUnexpected(WebCore::DOMCacheEngine::Error::CORP));
+            }
+
+            result.append(record);
```
WTFMove(record)

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+            result.append(record);
+        }
+
+        std::sort(result.begin(), result.end(), [&](const auto& a, const auto& b) {
```
s/[&]//

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+    
+    m_store->readRecords(targetRecordInfos, [options = WTFMove(options), callback = WTFMove(callback)](auto&& cacheStorageRecords) mutable {
+        Vector<WebCore::DOMCacheEngine::Record> result;
+        for (auto& cacheStorageRecord : cacheStorageRecords) {
```
Can we use WTF::map, or reserveInitialCapacity/uncheckedAppend

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+
+    int64_t spaceRequested = 0;
+    Vector<CacheStorageRecord> cacheStorageRecords;
+    for (auto&& record : records) {
```
WTF::map?

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+            record.request = WTFMove(existingRecord->request);
+            record.options = WTFMove(existingRecord->options);
+            record.referrer = WTFMove(existingRecord->referrer);
+            auto response = WebCore::ResourceResponse::fromCrossThreadData(WebCore::ResourceResponse::CrossThreadData { record.responseData });
```
Can we WTFMove record.responseData?

szewai replied:
hmm we need the `record.responseData` later for writing records to store (lines below), so probably not `WTFMove`

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,388 @@
+
+void CacheStorageCache::didPutOrRemoveRecords(Vector<uint64_t>&& targetIdentifiers, Vector<bool>&& results, WebCore::DOMCacheEngine::RecordIdentifiersCallback&& callback)
+{
+    bool succeeded = WTF::allOf(results, [](auto& result) {
```
It seems writeRecords should succeed or fail atomically, thus we should get a boolean here and not a Vector of bool.

szewai replied:
Yes the reason I didn't put it in `writeRecords` is it currently does not atomically succeed or fail -- it will not stop writing subsequent records when encountering an error. This is the behavior of current implementation.
I think we could make it return a boolean when we actually make `writeRecords` atomic (e.g. stopping writes and maybe rolling back to initial state before all writes).

youennf replied:
Atomic is probably what we wants, I would think getting the interface right and adding a comment in the cache storage disk store might be ok.

szewai replied:
Sure, I will change.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.h:
```diff
@@ -0,0 +1,71 @@
+    CacheStorageCache(CacheStorageManager&, const String& name, const String& uniqueName, const String& path, Ref<WorkQueue>&&);
+    WebCore::DOMCacheIdentifier identifier() const { return m_identifier; }
+    String name() const { return m_name; }
+    String uniqueName() const { return m_uniqueName; }
```
Why returning String instead of const String&?

szewai replied:
Will change.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+
+struct RecordHeader {
+    double insertionTime;
+    uint64_t size;
```
It would not hurt to set these to zero, ditto for responseBodySize and responseHeadersGuard

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+    : m_cacheName(cacheName)
+    , m_path(path)
+    , m_salt(valueOrDefault(FileSystem::readOrMakeSalt(saltFilePath())))
+    , m_queue(WTFMove(queue))
```
Maybe rename it m_callbackQueue?

szewai replied:
Okay.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+    if (buffer.isEmpty())
+        return std::nullopt;
+
+    auto fileData = Span { buffer.data(), buffer.size() };
```
Why not directly passing buffer and blobBuffer as a `Span<const uint8_t>` or maybe const Vector& at least?

szewai replied:
Will change.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+    return FileSystem::pathByAppendingComponent(blobsDirectoryPath(), blobFileName);
+}
+
+static std::optional<RecordMetaData> decodeRecordMetaData(Span<uint8_t> fileData)
```
Can we pass a `Span<const uint8_t>`?

szewai replied:
Will change.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+    return metaData;
+}
+
+static std::optional<RecordHeader> decodeRecordHeader(Span<uint8_t> headerData)
```
Ditto

szewai replied:
Sure

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+        Vector<CacheStorageRecord> result;
+        for (size_t index = 0; index < fileDatas.size(); ++index) {
+            if (auto record = readRecordFromFileData(fileDatas[index], blobDatas[index]))
+                result.append(WTFMove(*record));
```
Maybe add some logging if read is failing.
I wonder whether we should remove files if they are not decodable.

szewai replied:
I am not sure if we want to do write in a read operation so I didn't add it. I guess we could do a followup and add test if you feel like it's the right behavior.

youennf replied:
The write operation could be done asynchronously I guess.
In any case, adding some RELEASE_LOG_ERROR sounds like a good idea

szewai replied:
Will add loggings on read failure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+        Vector<Vector<uint8_t>> fileDatas;
+        Vector<Vector<uint8_t>> blobDatas;
+        auto partitionNames = FileSystem::listDirectory(recordsDirectory);
+        for (auto& paritionName : partitionNames) {
```
s/paritionName/partitionName/

szewai replied:
ah nice catch

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+            }
+        }
+
+        m_queue->dispatch([fileDatas = crossThreadCopy(fileDatas), blobDatas = crossThreadCopy(blobDatas), didReadRecordFiles = WTFMove(didReadRecordFiles)]() mutable {
```
WTFMove(fileDatas) and WTFMove(blobDatas) to allow optimised crossThreadCopy

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+{
+    Vector<String> recordFiles;
+    for (auto recordInfo : recordInfos)
+        recordFiles.append(recordFilePath(recordInfo.key));
```
WTF::map

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+        callback(WTFMove(result));
+    };
+
+    m_ioQueue->dispatch([this, protectedThis = Ref { *this }, recordFiles = crossThreadCopy(recordFiles), didReadRecordFiles = WTFMove(didReadRecordFiles)]() mutable {
```
WTFMove(recordFiles)

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+            blobDatas.append(WTFMove(blobData));
+        }
+
+        m_queue->dispatch([fileDatas = crossThreadCopy(fileDatas), blobDatas = crossThreadCopy(blobDatas), didReadRecordFiles = WTFMove(didReadRecordFiles)]() mutable {
```
WTFMove as well

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,520 @@
+        return recordFilePath(recordInfo.key);
+    });
+
+    m_ioQueue->dispatch([this, protectedThis = Ref { *this }, recordFiles = WTFMove(recordFiles), callback = WTFMove(callback)]() mutable {
```
WTFMove

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
```diff
@@ -0,0 +1,503 @@
+    if (m_isInitialized)
+        return true;
+
+    auto cachesList = readCachesList(m_path);
```
Should it be read asynchronously?

szewai replied:
I thought about this and figure all operations needs to be run after cachelist is read (`m_caches` is initialized), and cachelist should not be accessed as frequent as record files (e.g. client might open a cache and read/write multiple records). So I din't make it async.
We could do a followup if it's necessary.

youennf replied:
It would mean one client origin could delay another client origin since they would all use the same thread, correct?
If so, we should do this. A follow-up is probably fine but then let's keep track of this work, as well as the optimization work.

szewai replied:
Yes, I will add a comment here and file a bug.

szewai replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=249593.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
```diff
@@ -0,0 +1,503 @@
+        return m_caches[index]->open(WTFMove(callback));
+
+    m_caches.append(makeUnique<CacheStorageCache>(*this, name, createVersion4UUIDString(), m_path, m_queue.copyRef()));
+    bool written = writeCachesList(m_path, m_caches);
```
Ditto here

youennf commented on Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
```diff
@@ -64,58 +63,58 @@ void WebCacheStorageConnection::open(const WebCore::ClientOrigin& origin, const
         return;
 
-    connection().send(Messages::CacheStorageEngineConnection::Reference(cacheIdentifier), 0);
+    connection().send(Messages::NetworkStorageManager::Reference(cacheIdentifier), 0);
```
Some of these message names are no longer very meaningful given NetworkStorageManager is not CacheStorageSpecific.
We should probably update the names to make it clear they are CacheStorage related

szewai replied:
Sure, I will update the messages.

youennf reviewed (commented):
Another question, this patch is reducing bit the performances compared to the current code path. How are we supposed to make sure the new code path gets as good or better?
Should we have an easy way to resurrect/switch on the old code path?
Or are we close enough we are confident the optims we are thinking about will get us there?

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp:
```diff
@@ -0,0 +1,374 @@
+            record.request = WTFMove(existingRecord->request);
+            record.options = WTFMove(existingRecord->options);
+            record.referrer = WTFMove(existingRecord->referrer);
+            auto response = WebCore::ResourceResponse::fromCrossThreadData(record.responseData.copy());
```
We probably do not need to create a response just to get the Vary header. Let's make updateVaryHeaders take the Vary header value directly.

szewai replied:
Sounds good.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
```diff
@@ -0,0 +1,529 @@
+                result = false;
+        }
+
+        m_callbackQueue->dispatch([protectedThis = Ref { *this }, result, callback = WTFMove(callback)]() mutable {
```
Why protectedThis? If this is to make sure we do not delete this in the io queue, should we instead do protectedThis = WTFMove(protectedThis)?

szewai replied:
Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageMemoryStore.cpp:
```diff
@@ -0,0 +1,78 @@
+{
+    callback(WTF::map(m_records.values(), [](const auto& record) {
+        RELEASE_ASSERT(record);
+        return record->copy();
```
If the issue is that the CacheStorageMemoryStore can potentially be called from multiple threads, should we not do an isolated copy instead of a simple copy here?

szewai replied:
> should we not do an isolated copy instead of a simple copy here

Are you suggesting isolated copy or simple copy here?
`CacheStorageMemoryStore` is only accessed from one WorkQueue, and record does not have `AtomString`, so it seems safe to do a simple copy here.

youennf commented on Source/WebKit/NetworkProcess/storage/CacheStorageMemoryStore.cpp:
```diff
@@ -0,0 +1,78 @@
+            continue;
+        }
+        result.append(iterator->value->copy());
+    }
```
You can probably use WTF::map here

szewai replied:
Sure.

youennf reviewed (approved):
Some thoughts below.
As a further future optimization, I am wondering whether we should not consider sending IPC responses from whatever I/O thread we are using for reading/writing data.
This would remove hopping back to the IPC thread, which might improve reactivity a bit and may reduce memory copies.

szewai committed:
Manage CacheStorage by origin
https://bugs.webkit.org/show_bug.cgi?id=247207
rdar://problem/101968609

Reviewed by Youenn Fablet.

The OriginStorageManager class is designed to manage storage of a paritioned origin, and have exclusive access to origin
storage directory. This patch refactors DOMCache backend to use OriginStorageManager for CacheStorage management.

The old structure is:
NetworkSession - CacheStorage::Engine - CacheStorage::Caches - CacheStorage::Cache
The new structure is:
NetworkSession - NetworkStorageManager - OriginStorageManager - CacheStorageManager - CacheStorageCache

Here are the new classes:
CacheStorageManager: it manages cache storage of a partitioned origin and store caches.
CacheStorageCache: it represents a cache and keeps track of records (by record information) in memory.
CacheStorageStore: it represents backing store of cache; it is implemented by CacheStorageMemoryStore and
CacheStorageDiskStore.
CacheStorageRegistry: it keeps track of all CacheStorageCaches in a session, so NetworkStorageManager can find the
target of a CacheStorage IPC message quickly.

Here are some other changes made to complete the transition:
1. CacheStorage messages are sent to NetworkStorageManager instead of CacheStorage::Engine.
2. NetworkStorageManager handles CachesStorage messages and performs corresponding operations on its own WorkQueue.
Since WorkQueue is not tied to one thread, this patch use ResourceResponse::CrossThreadData instead of ResourceResponse
in CacheStorageRecord.
3. CacheStoragePersistentStore provides a simple file system backend for CacheStorage, instead of using
NetworkCache::Storage.
4. Representation string of CacheStorage will report origins that have CacheStorage data. It used to report origins of
existing CacheStorage::Caches (which manages cache storage of partitioned origin like CacheStorageManager), but having
CacheStorage::Caches does not mean the origin has CacheStorage data or the origin is actively performing CacheStorage
operations, and the state is subjective to change based on implementation (e.g. we might add/delete CacheStorage::Caches
proactively based on its activity). Therefore, let's use CacheStorage data as the truth, and update test expectations.
5. CacheStorage backend used to have multiple WorkQueues per partition origin, now it has one WorkQueue for all origins
(shared with other types) for message handling, and at most one queue per cache for I/O operation on cache records.

* LayoutTests/http/tests/cache-storage/cache-clearing-origin.https.html:
* LayoutTests/http/tests/cache-storage/cache-origins.https.html:
* LayoutTests/http/tests/cache-storage/cache-representation.https.html:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::overwriteEntireFile):
* Source/WTF/wtf/FileSystem.h:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::CrossThreadData::copy const):
* Source/WebCore/platform/network/ResourceResponseBase.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteDataForOrigins):
(WebKit::NetworkProcess::deleteAndRestrictWebsiteDataForRegistrableDomains):
(WebKit::NetworkProcess::registrableDomainsWithWebsiteData):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.cpp: Added.
(WebKit::computeKeyURL):
(WebKit::nextRecordIdentifier):
(WebKit::createStore):
(WebKit::CacheStorageCache::CacheStorageCache):
(WebKit::CacheStorageCache::manager):
(WebKit::CacheStorageCache::getSize):
(WebKit::CacheStorageCache::open):
(WebKit::toCacheStorageRecord):
(WebKit::CacheStorageCache::retrieveRecords):
(WebKit::CacheStorageCache::removeRecords):
(WebKit::CacheStorageCache::findExistingRecord):
(WebKit::CacheStorageCache::putRecords):
(WebKit::CacheStorageCache::putRecordsAfterQuotaCheck):
(WebKit::CacheStorageCache::putRecordsInStore):
(WebKit::CacheStorageCache::removeAllRecords):
(WebKit::CacheStorageCache::close):
* Source/WebKit/NetworkProcess/storage/CacheStorageCache.h: Added.
(WebKit::CacheStorageCache::identifier const):
(WebKit::CacheStorageCache::name const):
(WebKit::CacheStorageCache::uniqueName const):
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp: Added.
(WebKit::shouldStoreBodyAsBlob):
(WebKit::computeSHA1):
(WebKit::RecordMetaData::RecordMetaData):
(WebKit::CacheStorageDiskStore::create):
(WebKit::CacheStorageDiskStore::CacheStorageDiskStore):
(WebKit::CacheStorageDiskStore::versionDirectoryPath const):
(WebKit::CacheStorageDiskStore::saltFilePath const):
(WebKit::CacheStorageDiskStore::recordsDirectoryPath const):
(WebKit::CacheStorageDiskStore::recordFilePath const):
(WebKit::CacheStorageDiskStore::recordBlobFilePath const):
(WebKit::CacheStorageDiskStore::blobsDirectoryPath const):
(WebKit::CacheStorageDiskStore::blobFilePath const):
(WebKit::decodeRecordMetaData):
(WebKit::decodeRecordHeader):
(WebKit::CacheStorageDiskStore::readRecordFromFileData):
(WebKit::CacheStorageDiskStore::readAllRecords):
(WebKit::CacheStorageDiskStore::readRecords):
(WebKit::CacheStorageDiskStore::deleteRecords):
(WebKit::encodeRecordHeader):
(WebKit::encodeRecordBody):
(WebKit::encodeRecord):
(WebKit::CacheStorageDiskStore::writeRecords):
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.h: Added.
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp: Added.
(WebKit::nextUpdateNumber):
(WebKit::readCachesList):
(WebKit::writeCachesList):
(WebKit::readSizeFile):
(WebKit::writeSizeFile):
(WebKit::CacheStorageManager::cacheStorageSalt):
(WebKit::CacheStorageManager::cacheStorageOriginDirectory):
(WebKit::CacheStorageManager::originsOfCacheStorageData):
(WebKit::getDirectorySize):
(WebKit::CacheStorageManager::cacheStorageSize):
(WebKit::CacheStorageManager::hasCacheList):
(WebKit::CacheStorageManager::makeDirty):
(WebKit::CacheStorageManager::CacheStorageManager):
(WebKit::CacheStorageManager::~CacheStorageManager):
(WebKit::CacheStorageManager::initializeCaches):
(WebKit::CacheStorageManager::openCache):
(WebKit::CacheStorageManager::removeCache):
(WebKit::CacheStorageManager::allCaches):
(WebKit::CacheStorageManager::initializeCacheSize):
(WebKit::CacheStorageManager::finishInitializingSize):
(WebKit::CacheStorageManager::requestSpaceAfterInitializingSize):
(WebKit::CacheStorageManager::requestSpace):
(WebKit::CacheStorageManager::sizeIncreased):
(WebKit::CacheStorageManager::sizeDecreased):
(WebKit::CacheStorageManager::reference):
(WebKit::CacheStorageManager::dereference):
(WebKit::CacheStorageManager::connectionClosed):
(WebKit::CacheStorageManager::removeUnusedCache):
(WebKit::CacheStorageManager::hasDataInMemory):
(WebKit::CacheStorageManager::isActive):
(WebKit::CacheStorageManager::representationString):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.h: Added.
(WebKit::CacheStorageManager::salt const):
* Source/WebKit/NetworkProcess/storage/CacheStorageMemoryStore.cpp: Added.
(WebKit::CacheStorageMemoryStore::create):
(WebKit::CacheStorageMemoryStore::readAllRecords):
(WebKit::CacheStorageMemoryStore::readRecords):
(WebKit::CacheStorageMemoryStore::deleteRecords):
(WebKit::CacheStorageMemoryStore::writeRecords):
* Source/WebKit/NetworkProcess/storage/CacheStorageMemoryStore.h: Added.
* Source/WebKit/NetworkProcess/storage/CacheStorageRecord.h: Added.
(WebKit::CacheStorageRecordInformation::updateVaryHeaders):
(WebKit::CacheStorageRecord::CacheStorageRecord):
(WebKit::CacheStorageRecord::copy const):
* Source/WebKit/NetworkProcess/storage/CacheStorageRegistry.cpp: Added.
(WebKit::CacheStorageRegistry::registerCache):
(WebKit::CacheStorageRegistry::unregisterCache):
(WebKit::CacheStorageRegistry::cache):
* Source/WebKit/NetworkProcess/storage/CacheStorageRegistry.h: Added.
* Source/WebKit/NetworkProcess/storage/CacheStorageStore.h: Added.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::canHandleTypes):
(WebKit::NetworkStorageManager::originStorageManager):
(WebKit::NetworkStorageManager::getAllOrigins):
(WebKit::NetworkStorageManager::cacheStorageOpenCache):
(WebKit::NetworkStorageManager::cacheStorageRemoveCache):
(WebKit::NetworkStorageManager::cacheStorageAllCaches):
(WebKit::NetworkStorageManager::cacheStorageReference):
(WebKit::NetworkStorageManager::cacheStorageDereference):
(WebKit::NetworkStorageManager::cacheStorageRetrieveRecords):
(WebKit::NetworkStorageManager::cacheStorageRemoveRecords):
(WebKit::NetworkStorageManager::cacheStoragePutRecords):
(WebKit::NetworkStorageManager::cacheStorageClearMemoryRepresentation):
(WebKit::NetworkStorageManager::cacheStorageRepresentation):
(WebKit::readOriginFromFile): Deleted.
(WebKit::writeOriginToFile): Deleted.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::StorageBucket::existingCacheStorageManager):
(WebKit::OriginStorageManager::StorageBucket::StorageBucket):
(WebKit::OriginStorageManager::StorageBucket::connectionClosed):
(WebKit::OriginStorageManager::StorageBucket::toStorageType const):
(WebKit::OriginStorageManager::StorageBucket::toStorageIdentifier const):
(WebKit::OriginStorageManager::StorageBucket::cacheStorageManager):
(WebKit::OriginStorageManager::StorageBucket::isActive const):
(WebKit::OriginStorageManager::StorageBucket::isEmpty):
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInListFromMemory):
(WebKit::OriginStorageManager::StorageBucket::fetchDataTypesInListFromDisk):
(WebKit::OriginStorageManager::StorageBucket::deleteData):
(WebKit::OriginStorageManager::StorageBucket::deleteCacheStorageData):
(WebKit::OriginStorageManager::StorageBucket::deleteEmptyDirectory):
(WebKit::OriginStorageManager::StorageBucket::resolvedPath):
(WebKit::OriginStorageManager::StorageBucket::closeCacheStorageManager):
(WebKit::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
(WebKit::OriginStorageManager::defaultBucket):
(WebKit::OriginStorageManager::existingCacheStorageManager):
(WebKit::OriginStorageManager::cacheStorageManager):
(WebKit::OriginStorageManager::closeCacheStorageManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/NetworkProcess/storage/StorageUtilities.h: Added.
(WebKit::readOriginFromFile):
(WebKit::writeOriginToFile):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp:
(WebKit::WebCacheStorageConnection::open):
(WebKit::WebCacheStorageConnection::remove):
(WebKit::WebCacheStorageConnection::retrieveCaches):
(WebKit::WebCacheStorageConnection::retrieveRecords):
(WebKit::WebCacheStorageConnection::batchDeleteOperation):
(WebKit::WebCacheStorageConnection::batchPutOperation):
(WebKit::WebCacheStorageConnection::reference):
(WebKit::WebCacheStorageConnection::dereference):
(WebKit::WebCacheStorageConnection::clearMemoryRepresentation):
(WebKit::WebCacheStorageConnection::engineRepresentation):

Canonical link: https://commits.webkit.org/258235@main
